envoy/config/xds_config_tracer.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/optref.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+#include "google/rpc/status.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+// The status of processing Delta/DiscoveryResponse.
+enum TraceState {
+  // Successfully got the resources or message.
+  RECEIVE = 0,
+  // Successfully ingested the resouces.
+  INGESTED = 1,
+  // Failed to apply the resources.
+  FAILED = 2,
+};
+
+struct TraceDetails {
+  TraceDetails(const TraceState state) : state_(state){};
+  TraceDetails(const TraceState state, const ::google::rpc::Status error_detail)
+      : state_(state), error_detail_(error_detail){};
+  const TraceState state_;
+  ::google::rpc::Status error_detail_;
+};
+
+/**
+ * An interface for hooking into xDS update events to provide the ablility to use some
+ * external logger or processor in xDS update.
+ *
+ * Instance of this interface get invoked on the main Envoy thread. Thus, it is important
+ * for implementations of this interface to not execute any blocking operations on the same
+ * thread.
+ */
+class XdsConfigTracer {
+public:
+  virtual ~XdsConfigTracer() = default;
+
+  /**
+   * Log the decoded SotW xDS resources that are about to be ingested.
+   * @param type_url The type url of xDS message.
+   * @param resources List of decoded resources that reflect the latest state.
+   * @param details The log point state and details.
+   */
+  virtual void log(const absl::string_view type_url,
```

Comment:
Why is this "log" and not just a generic hook (e.g. "onConfigIngestion")  in the xDS ingestion lifecycle? Is there some documentation on when this is called relative to the xDS delegate, etc?

Reply:
Thanks for taking a look! And updated the PR. There is a high level doc internally, but doesn't include the details related to this PR. I will add more comments in the PR and here. The initial purpose is to add some log point and persist the log to adapt the external package, therefore `log` is used.

For both SotW and Delta response message, we will have three log points: RCEIVE, INGESTED, FAILED. 
For SotW, the first log method is used twice, one is at the time when the resource is successfully parsed and before the ingestion. The other one is when the resource is successfully applied to Envoy and are about to be ACK'ed. The TraceDetails implies the status of ingestion process, e.g., TraceState::INGESTED. The log point TraceState::INGESTED is the same place as xDS delegate, like we can use `onConfigIngestion`. 

Similar to Delta, the log point is distinguished by the TraceState: RCEIVE, INGESTED, FAILED. The extension is more like a logger, but it could be used by other purpose. This is my initial thinking, happy to learn any thoughts. Thank you!

Reply:
I think what @htuch implies is that this can be generalized to some other use-cases that would want to know the state of resources during the config ingestion process. In other words, not only for tracing/logging, but more generic.

Specifically, the callback methods can be called `resourcesProcessingUpdate`/`configProcessingUpdate` and `TraceState` could be renamed to `ProcessingState` (as it is already under the `Envoy::Config` namespace).

Reply:
Yep - I don't think we need any deep changes, but just to model this hook as part of the xDS configuration ingestion lifecycle, rather than being super opinionated about this being a log or trace thingy (maybe we just want to bump a stat, or post an event to some Dispatcher as part of on-demand xDS etc.)

Reply:
Gotcha, generalization makes sense. Will change & thank you all!